### PR TITLE
Put the WorkflowInfo public class back, but marked as obsolete

### LINKED ISSUE #11053: Binary modules that worked in PowerShell 6 may not work in 7
@iSazonov and @SteveL-MSFT I think we made a mistake by removing the `WorkflowInfo` type in PR #9638. I know PR #6708 changed the type from public to internal, and that was probably a mistake as well.

I was just testing my FormatPx module in PowerShell 7, since `Out-Printer` is now added back into PowerShell, but the module won't load. It fails with the following error:

> Import-Module: Could not load type 'System.Management.Automation.WorkflowInfo' from assembly 'System.Management.Automation, Version=7.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35'."

I get the same error if I try to import DebugPx into PowerShell 7, but DebugPx loads just fine in PowerShell 6.2 (FormatPx does not, because there is no `Out-Printer` command, and it derives its `Out-Printer` proxy from that command).

Both DebugPx and FormatPx have binary modules as part of their package.

I am not referencing `WorkflowInfo` directly from either of those projects, so that error must be coming from the PowerShell components that I do reference.

Testing this out locally, I have found that if I put an empty `public WorkflowInfo` class back in PowerShell 7 (not derived from anything, with no constructors or members), suddenly my DebugPx module  loads fine again.

FormatPx loads as well, but doesn't work (null reference exception somewhere), but so I need to figure that one out. That issue aside, it seems we need the `WorkflowInfo` class to persist for backwards compatibility support. Mind if I put it back in a PR?

# Steps to reproduce

```powershell
Install-Module DebugPx
Import-Module DebugPx
```

# Expected behavior

The module should load without error.

# Actual behavior

```none
Import-Module: Could not load type 'System.Management.Automation.WorkflowInfo' from assembly 'System.Management.Automation, Version=7.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35'."
```

# Environment data

```none
Name                           Value
----                           -----
PSVersion                      7.0.0-preview.5
PSEdition                      Core
GitCommitId                    7.0.0-preview.5-64-ge185f89591ac5fa5286bef60d087d852873df2be
OS                             Microsoft Windows 10.0.17763
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### PR DESCRIPTION
**Pull request by KirkMunro**
# PR Summary

Add an empty, obsolete `WorkflowInfo` public class back into PowerShell.

## PR Context

Fixes #11053.

## PR Checklist

- [X] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [X] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [X] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [X] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [X] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [X] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

**Review by iSazonov — APPROVED** [review]

**Review by daxian-dbw — APPROVED** [review]

File: src/System.Management.Automation/engine/WorkflowInfo.cs [hunk 1]
```diff
@@ -0,0 +1,18 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+namespace System.Management.Automation
+{
+    /// <summary>
+    /// Deprecated. Do not use.
+    /// </summary>
+    /// <remarks>
+    /// This class must not be deleted. Existing modules depend on this public
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/WorkflowInfo.cs]:
```suggestion
    /// This class must not be deleted. Existing modules may depend on this public
```

**Comment by SeeminglyScience** [comment]
Any reason not to include the original subclass and members?  Won't this still throw on JIT of a method that references one of these members?

**Comment by iSazonov** [comment]
@SeeminglyScience This would mean that users would expect they (subclass and members) to _work_, but it is not. 
I think the current compromise is good.

**Comment by SeeminglyScience** [comment]
> @SeeminglyScience This would mean that users would expect they (subclass and members) to _work_, but it is not.
> I think the current compromise is good.

Take this example:

```csharp
string GetRealName(CommandInfo command)
{
    return command switch
    {
        WorkflowInfo workflow => workflow.Name,
        FunctionInfo function => function.Name,
        _ => string.Empty,
    };
}
```

There isn't going to be any workflows coming into this method because core doesn't support it.  The property can throw `NotSupportedException`, it doesn't really matter because that code path will never be hit.  But just by referencing the property this method can no longer be JIT'd (if I'm remembering correctly).

**Comment by iSazonov** [comment]
@SeeminglyScience You could add a method call... It has no end. Really the PR addresses an issue reported by Kirk, no more. If somebody point another project which we must save we could consider this. Although this does not make sense because it has already been three years since we removed workflow support. Now it's time to remove all extra links.

**Comment by SeeminglyScience** [comment]
> @SeeminglyScience You could add a method call... It has no end.

Not sure what you mean. Keeping the layout of the class would be the end and not really take any additional effort or complexity.

> Really the PR addresses an issue reported by Kirk, no more. If somebody point another project which we must save we could consider this. Although this does not make sense because it has already been three years since we removed workflow support. Now it's time to remove all extra links.

I'd be more inclined to agree if you were suggesting that the class shouldn't be added back at all.  I don't see what is lost by including the members as well. I get that there's no reported issue with them *yet*, but decorating them with `Obsolete` solves any implied support issues and there's no real downside if you're already including the class.

**Comment by iSazonov** [comment]
@SeeminglyScience My thoughts is simple. If anybody creates an instance of the class he expects that the class works but really it does not work and it is good to throw, otherwise it can lead to unpredictable behavior.

**Comment by SeeminglyScience** [comment]
> @SeeminglyScience My thoughts is simple. If anybody creates an instance of the class he expects that the class works but really it does not work and it is good to throw, otherwise it can lead to unpredictable behavior.

Yeah for sure, throw `NotSupportedException` at runtime if that code path is actually hit, 100% agree.   Really even the constructors should throw.  The point in adding this back is only for assembly load/JIT time backwards compatibility right?  If someone is trying to actually instansiate an instance of `WorkflowInfo`, that should definitely throw as well. My issue is only with JIT time exceptions, they're way harder to troubleshoot and throw even when someone is just trying to account for the *possibility* of being passed a `WorkflowInfo` object.

**Comment by KirkMunro** [comment]
> Any reason not to include the original subclass and members? Won't this still throw on JIT of a method that references one of these members?

Not really. It simply came down to time and need for me. I put back what I had time to put back, and addressed the specific issue I came across. I have a trip next week that is naturally making me frantic because as usual I need to do all the things before that trip since trips really only shift work rather than remove work from your plate, so I just did a quick fix here because that's what I had time to do. If we want more than what I did, that's fine by me. I'm too busy to think about it myself though, at least until I'm back from my trip the week after next.

**Comment by joeyaiello** [comment]
We discussed this at length within the @PowerShell/powershell-committee and went back and forth a few times:

It's been our philosophy for a long time that we should make a "best effort" to load .NET Framework / WinPS compiled modules in .NET/PS Core. E.g. if 9/10 cmdlets in a module would work just fine at runtime, but the 1 cmdlet contains a reference to a missing type, it gives me heartburn that we block import altogether. This is particularly relevant with regards to modules that are not and will not be maintained but that still can have value in (mostly) working with PS7. 

That being said, we also don't want to start adding these things back on an ad hoc basis. We removed a ton of stuff over the course of 6.x, and we certainly shouldn't re-introduce just this one type if we decide that it's the right thing. The effort level to enumerate, analyze, decide on, and reintroduce all those types would be significant, and may cause other unforeseen problems. Therefore, we'd rather wait until we see this problem more often in the wild--and particularly around modules that are not well maintained--before it makes sense to make that investment. 

Also, we all agree that we wouldn't ever want to add something like this back to .NET Standard, as we're trying to get as close to of a guarantee of "it will work on both editions" as is possible at compile time. Furthermore, if a module is being recompiled, then it means it's being actively maintained, and the usage of that type can be removed.

**Comment by KirkMunro** [comment]
@joeyaiello I think you made the wrong call on this. I chose to apply the PR because that was something I could do today. Obviously if I could have just fixed this in my modules I would have already, but releasing an update to my modules is not something I can do until I have a new code signing certificate.

> That being said, we also don't want to start adding these things back on an ad hoc basis. We removed a ton of stuff over the course of 6.x, and we certainly shouldn't re-introduce just this one type if we decide that it's the right thing.

This is a new change in PowerShell 7. It was not in PowerShell 6.x. Ergo it isn't re-introducing something, it's just undoing the removal of that thing before PowerShell 7 is released.

> Therefore, we'd rather wait until we see this problem more often in the wild--and particularly around modules that are not well maintained--before it makes sense to make that investment.

This is a specific issue with a problem in the wild, for a change that has not yet been included in anything other than a pre-release. It's also a problem with a module that is used by other modules, so the impact isn't just localized to one thing. I don't see where "investment" comes to play here given that this is pre-release.

> Furthermore, if a module is being recompiled, then it means it's being actively maintained, and the usage of that type can be removed.

I only compiled the module again to identify where the issue was coming from. While I want to actively maintain my modules, I am not in a position to code-sign them yet, and as such, they are sitting idle until I can code-sign them again. This is something I hope to be in a position to do again in the next few months, but it is not something I can do today. That doesn't qualify for "actively maintained" in my book.

**Comment by iSazonov** [comment]
I think that this will not worsen PowerShell if we leave this class for another six months with obsolete attribute.

**Pull request closed without merging**